gym/envs/classic_control/cartpole.py

```diff
@@ -60,10 +60,10 @@ def step(self, action):
         temp = (force + self.polemass_length * theta_dot * theta_dot * sintheta) / self.total_mass
         thetaacc = (self.gravity * sintheta - costheta* temp) / (self.length * (4.0/3.0 - self.masspole * costheta * costheta / self.total_mass))
         xacc  = temp - self.polemass_length * thetaacc * costheta / self.total_mass
-        x  = x + self.tau * x_dot
         x_dot = x_dot + self.tau * xacc
-        theta = theta + self.tau * theta_dot
+        x  = x + self.tau * x_dot
         theta_dot = theta_dot + self.tau * thetaacc
+        theta = theta + self.tau * theta_dot
```

Comment:
why should it be in this new order? To me it would seem that Euler-method-like step should be
x_n+1 = x_n + tau * (xdot_n)
xdot_n+1 = xdot_n + (acc_n) 
etc which corresponded to the old version... Are there some stability issues?

Reply:
I looked at the permalink and I think even the original had this bug.  It is the order of operations in calculus.  You need to integrate acceleration to get velocity, and integrate velocity to get position.  If you switch up the order and do acceleration -> position -> velocity, the position will always be one timestamp behind because it will have used the previous velocity rather than the current velocity.

So for example if I gave you a math problem where you had the acceleration given you would have to calculate the velocity first to get the position, it's the same thing here.  The reason it works is because on line 57 you are reading in the previous velocity.

https://github.com/openai/gym/blob/293eea787a662f501b0e4aab512d3769e830ece2/gym/envs/classic_control/cartpole.py#L57

Reply:
The integration that the code does now is called semi-implicit / symplectic euler integration, and is much more stable than euler integration. Most game physics simulators use this method. This isn't a bug.

Reply:
We should probably comment this on the code

Reply:
Oops, after looking more closely, it's just normal Euler integration as Peter said, and it's correct.

Reply:
I am pretty sure they are kinematic equations #1 and #2 with some simplifications.  If they are, the order of the lines should be switched.

https://www.khanacademy.org/science/physics/one-dimensional-motion/kinematic-formulas/a/what-are-the-kinematic-formulas

Reply:
An airplane accelerates down a runway at 3.20 m/s2 for 32.8 s until is finally lifts off the ground. Determine the distance traveled before takeoff.

If you take the above example, you would have to calculate the velocity before the position using the kinematic equations.  The process is the same as what is being used in the code.

Reply:
Per discussion with @joschu - both versions are correct. The old one is the vanilla Euler, and the suggested is semi-implicit Euler. While the new one is more stable, we'd rather not change the behavior of the existing environment. If you'd still like to use semi-implicit Euler, could you add a separate environment flag that turns it on (i.e. by default it is off, and a flag can turn it on)?